---
title: "Material Comparison"
author: "Todd Testerman"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(phyloseq, BiocManager, qiime2R, tidyverse, viridis, pheatmap, ggplot2, vegan, patchwork, ggpubr, ranacapa, nlme, compositions, readr, devtools, DESeq2, microViz, ANCOMBC, microbiome, GUniFrac, rstatix, corncob, scales, MicEco, btools, dplyr, tibble, pairwiseAdonis, metagMisc, biomehorizon, devEMF, cowplot, microbiomeutilities, UpSetR, MicrobiotaProcess, ComplexUpset, randomForest, doParallel, NetCoMi)

set.seed(816)
permutations = 9999
```


Import

```{r}
phy = qza_to_phyloseq(features = "table-all-runs.qza", tree = "phylogeny/rooted_tree.qza", taxonomy = "taxonomy-sv-all-runs.qza", metadata = "mapping_file_all_runs.txt",tmp = "/var/folders/67/6dbph7ys1d3240bqc4wgn7380000gp/T/")
ntaxa(phy)
nsamples(phy)

#tax_fix_interactive(phy)
#adjust tax table for use in microviz
phy = phy%>%
 tax_fix(
  min_length = 3,
  unknowns = c("","Incertae_Sedis", "uncultured", "Unknown_Family", "Unknown_Family Family"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )
#remove prefix
phy@tax_table = gsub("d__", "", phy@tax_table)
#treat Maturity Dayss as factors to set order as early, middle, late
phy@sam_data$Maturity_Days = factor(phy@sam_data$Maturity_Days, levels = c("9", "23", "38", "53", "65", "80"))
```



```{r}
#subset to only include day 23 wall swabs for comparison between stainless steel tank walls and concrete walls
phy_ws_23d = subset_samples(phy, Maturity_Days == "23")
phy_ws_23d = subset_samples(phy_ws_23d, Sample_Type == "Wall")
phy_ws_23d_rarefy = rarefy_even_depth(phy_ws_23d, 8000)
nsamples(phy_ws_23d)
```

```{r}
genus_palette = distinct_palette(n = 41)
names(genus_palette) = tax_top(phy_ws_23d, n = 41, by = sum, rank = "Genus")
names(genus_palette)[42] = "Other"

family_palette = distinct_palette(n = 41)
names(family_palette) = tax_top(phy_ws_23d, n = 41, by = sum, rank = "Family")
names(family_palette)[42] = "Other"

class_palette = distinct_palette(n = 41)
names(class_palette) = tax_top(phy_ws_23d, n = 41, by = sum, rank = "Class")
names(class_palette)[42] = "Other"

#set figure theme
theme_settings = theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), strip.text.x = element_text(size = 8, face = "bold"), axis.text.y = element_text(size = 8, face = "bold"), axis.title.y = element_text(size = 12, face = "bold"), legend.key.size = unit(0.2, 'cm'))
```




Stainless vs Concrete Painted Walls


```{r, fig.width= 10, fig.height=12} 
phy_ws_23d_filter = tax_filter(phy_ws_23d, min_prevalence = 2)

#ord_explore(phy_ws_23d_filter)

#rarefied
#ord_explore(phy_ws_23d_rarefy)

#bray
wall_swabs_bray = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "bray") %>%
 ord_calc(
  method = "NMDS"
 ) %>% 
 ord_plot(
  axes = c(1, 2),
  colour = "Material", fill = "Material",
  shape = "circle", alpha = 0.5,
  size = 2, auto_caption = NA
 ) +
 ggplot2::stat_ellipse(
  ggplot2::aes(colour = Material)
 ) + 
  labs(color = "Material", fill = "Material")
wall_swabs_bray

bray_perm_results = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "bray") %>%
 dist_permanova(
    variables = "Material",
    n_processes = 10,
  ) 
bray_perm_results
write.table(bray_perm_results@permanova, "bray_perm_results_material.csv", sep = ",")
pairwise_bray_perm_results = pairwise.adonis(x = otu_get(phy_ws_23d_rarefy@otu_table), factors = sample_data(phy_ws_23d_rarefy)$Material, sim.method = "bray", p.adjust.m = "holm", perm = 999999)
write.table(pairwise_bray_perm_results, "pairwise_bray_perm_results_material.csv", sep = ",")
#Bray beta disp
bray_bdisp_results = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "bray") %>%
 dist_bdisp(
    variables = "Material",
  ) 
bray_bdisp_results@bdisp
#jaccard
wall_swabs_jacc = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "jaccard", binary = T) %>%
 ord_calc(
  method = "NMDS"
 ) %>% 
 ord_plot(
  axes = c(1, 2),
  colour = "Material", fill = "Material",
  shape = "circle", alpha = 0.5,
  size = 2, auto_caption = NA
 ) +
 ggplot2::stat_ellipse(
  ggplot2::aes(colour = Material)
 ) + 
  labs(color = "Material", fill = "Material")
wall_swabs_jacc


jacc_perm_results = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "jaccard", binary = T) %>%
 dist_permanova(
    variables = "Material",
    n_processes = 10,
  ) 
jacc_perm_results
write.table(jacc_perm_results@permanova, "jacc_perm_results_material.csv", sep = ",")
pairwise_jacc_perm_results = pairwise.adonis(x = otu_get(phy_ws_23d_rarefy@otu_table), factors = sample_data(phy_ws_23d_rarefy)$Material, sim.method = "jaccard", p.adjust.m = "holm", perm = 999999)
write.table(pairwise_jacc_perm_results, "pairwise_jacc_perm_results_material.csv", sep = ",")


#Jaccard beta disp
jacc_bdisp_results = phy_ws_23d_rarefy %>%
 tax_transform(rank = "unique", trans = "identity") %>%
 dist_calc(dist = "jaccard") %>%
 dist_bdisp(
    variables = "Material",
  ) 
jacc_bdisp_results@bdisp
#weighted unifrac
wall_swabs_wunifrac_calc = phy_ws_23d_rarefy %>% 
  dist_calc(dist = "wunifrac")

wall_swabs_wunifrac = wall_swabs_wunifrac_calc %>%
 ord_calc(
  method = "NMDS"
 ) %>% 
 ord_plot(
  axes = c(1, 2),
  colour = "Material", fill = "Material",
  shape = "circle", alpha = 0.5,
  size = 2, auto_caption = NA
 ) +
 ggplot2::stat_ellipse(
  ggplot2::aes(colour = Material)
 ) + 
  labs(color = "Material", fill = "Material")
wall_swabs_wunifrac


wunifrac_perm_results = wall_swabs_wunifrac_calc %>%
 dist_permanova(
    variables = "Material",
    n_processes = 10,
  ) 
wunifrac_perm_results
write.table(wunifrac_perm_results@permanova, "wunifrac_perm_results_material.csv", sep = ",")
pairwise_wunifrac_perm_results = pairwise.adonis(x = wall_swabs_wunifrac_calc@dist, factors = sample_data(phy_ws_23d_rarefy)$Material, sim.method = "wunifrac", p.adjust.m = "holm", perm = 999999)
write.table(pairwise_wunifrac_perm_results, "pairwise_wunifrac_perm_results_material.csv", sep = ",")

#Wunifrac beta disp
wunifrac_bdisp_results = wall_swabs_wunifrac_calc %>%
 dist_bdisp(
    variables = "Material",
  ) 
wunifrac_bdisp_results@bdisp
#This is the only significant bdisp result - dispersion is not homogenous between early and late wall swabs for weighted unifrac metric

#unweighted unifrac
wall_swabs_unifrac_calc = phy_ws_23d_rarefy %>% 
  dist_calc(dist = "unifrac")

wall_swabs_unifrac = wall_swabs_unifrac_calc %>%
 ord_calc(
  method = "NMDS"
 ) %>% 
 ord_plot(
  axes = c(1, 2),
  colour = "Material", fill = "Material",
  shape = "circle", alpha = 0.5,
  size = 2, auto_caption = NA
 ) +
 ggplot2::stat_ellipse(
  ggplot2::aes(colour = Material)
 ) + 
  labs(color = "Material", fill = "Material")
wall_swabs_unifrac


unifrac_perm_results = wall_swabs_unifrac_calc %>%
 dist_permanova(
    variables = "Material",
    n_processes = 10,
  ) 
unifrac_perm_results
write.table(unifrac_perm_results@permanova, "unifrac_perm_results_material.csv", sep = ",")
pairwise_unifrac_perm_results = pairwise.adonis(x = wall_swabs_unifrac_calc@dist, factors = sample_data(phy_ws_23d_rarefy)$Material, sim.method = "unifrac", p.adjust.m = "holm", perm = 999999)
write.table(pairwise_unifrac_perm_results, "pairwise_unifrac_perm_results_material.csv", sep = ",")

#Unifrac beta disp
unifrac_bdisp_results = wall_swabs_unifrac_calc %>%
 dist_bdisp(
    variables = "Material",
  ) 
unifrac_bdisp_results@bdisp
#Patchwork for Figure 2 - 4 metrics NMDS
Figure_2 = wall_swabs_bray + wall_swabs_jacc + wall_swabs_wunifrac + wall_swabs_unifrac + plot_annotation(tag_levels = "A") + plot_layout(guides = "collect")
Figure_2

ggsave("Figures_all_Material/Figure_2_beta_div.png", width= 10, height=12)
ggsave("Figures_all_Material/Figure_2_beta_div.tiff", dpi = 300)

save_plot("Figures_all_Material/Figure_2_beta_div_cow.tiff", Figure_2, base_height = 6, ncol = 2, nrow = 2)

#SFig PCA arrows plus iris
phy_ws_23d_filter@sam_data$Material = factor(phy_ws_23d_filter@sam_data$Material, levels = c("9", "23", "38", "53", "65", "80"))

PCA_wall_swabs =  phy_ws_23d %>%
 tax_transform(rank = "Genus", trans = "clr") %>%
 ord_calc(
  method = "auto"
 ) %>% 
 ord_plot(
  axes = c(1, 2),
  plot_taxa = 1:10,
  colour = "Material", fill = "Material",
  shape = "circle", alpha = 0.5,
  size = 2
 )

PCA_wall_swabs = PCA_wall_swabs +
 ggplot2::stat_ellipse(
  ggplot2::aes(colour = Material)) +
  labs(color = "Material", fill = "Material") +
  theme(axis.ticks.x = element_blank(), strip.text.x = element_text(size = 8, face = "bold"), axis.text.x = element_text(size = 8, face = "bold"), axis.text.y = element_text(size = 8, face = "bold"), axis.title.x = element_text(size = 12, face = "bold"), axis.title.y = element_text(size = 12, face = "bold"), legend.key.size = unit(0.2, 'cm'))
  
iris_plot = phy_ws_23d_filter %>% 
  tax_transform(rank = "Genus", trans = "clr") %>%
 ord_calc(
  method = "auto") %>% 
  ord_plot_iris(tax_level = "Genus", ord_plot = "none", anno_colour = "Material", n_taxa = 20, palette = genus_palette) +
  guides(color = "none")

SFig_PCA_arrows_iris_plot = PCA_wall_swabs / iris_plot + plot_annotation(tag_levels = "A") + plot_layout(guides = "collect")
#ggsave("Figures_all_Material/SFig_PCA_arrows_iris_plot.png", width= 10, height=12)
#ggsave("Figures_all_Material/SFig_PCA_arrows_iris_plot.tiff", dpi = 300)

#save_plot("Figures_all_Material/SFig_PCA_arrows_iris_cowplot.tiff", SFig_PCA_arrows_iris_plot, base_height = 8)
```

Alpha Diversity

```{r}
#faith
faith_all_groups_ws = estimate_pd(phy_ws_23d_rarefy)
faith_all_groups_ws = rownames_to_column(faith_all_groups_ws, "Sample")
psmelt_ws_rarefy = psmelt(phy_ws_23d_rarefy)
psmelt_ws_rarefy = psmelt_ws_rarefy %>% distinct(Sample, .keep_all = T)
merged_for_faith_ws = merge(psmelt_ws_rarefy, faith_all_groups_ws, by = "Sample")
merged_for_faith_ws = as_data_frame(merged_for_faith_ws)

faith_ws = ggplot(merged_for_faith_ws, aes(y=PD, x=Material, color=Material)) + geom_boxplot(show.legend = T, outlier.size = 0.2, lwd = 0.25) + geom_point(size = 0.2) 
faith_ws = faith_ws + theme_bw() + theme(strip.background = element_blank(), strip.text.x = element_blank(), axis.text.x = element_blank(), axis.title.x = element_blank()) + labs(color = "Material") + ylab("Faith's PD Index Value")
faith_ws

#normality test
merged_for_faith_ws %>% group_by(Material) %>% shapiro_test(PD)
#normal distribution for all groups -------> use parametric test
faith_ttest_all_groups_ws = merged_for_faith_ws %>% t_test(PD ~ Material)
faith_ttest_all_groups_ws
#export stats, if desired
#write.table(faith_ttest_all_groups_ws, "faith_ttest_ws.csv", sep = ",")

#Shannon
shannon_ws = plot_richness(phy_ws_23d_rarefy, x = "Material", measures = "Shannon", color = "Material")
shannon_ws = shannon_ws + geom_boxplot(data = shannon_ws$data, show.legend = T, outlier.size = 0.2, lwd = 0.25) + geom_point(size = 0.2)
shannon_ws$layers = shannon_ws$layers[-1]
shannon_ws = shannon_ws + theme_bw() + theme(strip.background = element_blank(), strip.text.x = element_blank(), axis.text.x = element_blank(), axis.title.x = element_blank()) + labs(color = "Material") + ylab("Shannon Diversity Index Value")
shannon_ws

#normality test
shannon_data_ws = shannon_ws$data
colnames(shannon_data_ws)[35] = "Shannon_Index"
shannon_data_ws %>% group_by(Material) %>% shapiro_test(Shannon_Index)
#non-normal distribution for 2/3 groups -------> use non-parametric test
shannon_wilcox_all_groups_ws = shannon_data_ws %>% wilcox_test(Shannon_Index ~ Material)
shannon_wilcox_all_groups_ws
#export stats, if desired
#write.table(shannon_wilcox_all_groups_ws, "shannon_wilcox_ws.csv", sep = ",")

#Observed ASVs
observed_ASV_ws = plot_richness(phy_ws_23d_rarefy, x = "Material", measures = "Observed", color = "Material")
observed_ASV_ws = observed_ASV_ws + geom_boxplot(data = observed_ASV_ws$data, show.legend = T, outlier.size = 0.2, lwd = 0.25) + geom_point(size = 0.2)
observed_ASV_ws$layers = observed_ASV_ws$layers[-1]
observed_ASV_ws = observed_ASV_ws + theme_bw() + theme(strip.background = element_blank(), strip.text.x = element_blank(), axis.text.x = element_blank(), axis.title.x = element_blank()) + labs(color = "Material") + ylab("Observed ASVs")
observed_ASV_ws

#normality test
observed_ASV_data_ws = observed_ASV_ws$data
colnames(observed_ASV_data_ws)[35] = "Observed_ASVs"
observed_ASV_data_ws %>% group_by(Material) %>% shapiro_test(Observed_ASVs)
#non-normal distribution for 2/3 groups -------> use non-parametric test
observed_ASV_wilcox_all_groups_ws = observed_ASV_data_ws %>% wilcox_test(Observed_ASVs ~ Material)
observed_ASV_wilcox_all_groups_ws
#export stats, if desired
#write.table(observed_ASV_wilcox_all_groups_ws, "observed_ASV_wilcox_ws.csv", sep = ",")

#Figure 3
Figure_3_alpha_div = observed_ASV_ws + shannon_ws + faith_ws
Figure_3_alpha_div = Figure_3_alpha_div + plot_annotation(tag_levels = 'A') + plot_layout(guides = 'collect')
Figure_3_alpha_div

#ggsave("Figures_all_Material/Figure_3_alpha_div.png")
#ggsave("Figures_all_Material/Figure_3_alpha_div.tiff", dpi = 300)

#save_plot("Figures_all_Material/Figure_3_alpha_div_cow.tiff", Figure_3_alpha_div, base_height = 8)
```



```{r, fig.width= 10, fig.height=12} 
#ANCOM-BC

phy_ws_23d_filter_class = aggregate_taxa(phy_ws_23d, "Class")
phy_ws_23d_filter_class_filtered = tax_filter(phy_ws_23d_filter_class, min_prevalence = 1/10)

#Material

ancom_material_class = ancombc(phy_ws_23d_filter_class_filtered, group = "Material", formula = "Material", neg_lb = F)
res_material_class = ancom_material_class$res
as.data.frame(res_material_class)
#Waterfall plot
df_fig1_class_material = res_material_class$lfc %>%
  as.data.frame() %>%
  filter(res_material_class$diff_abn$MaterialSteel) %>%
  rename(taxon_id = taxon)  # Rename taxon for consistency

df_fig2_class = data.frame(res_material_class$se, check.names = FALSE) %>%
  rownames_to_column("taxon_id") %>%
  filter(res_material_class$diff_abn$MaterialSteel)

colnames(df_fig2_class)[-1] = paste0(colnames(df_fig2_class)[-1], "SD")

df_fig_class = df_fig1_class_material %>% left_join(df_fig2_class, by = "taxon_id") %>%
  transmute(taxon_id, MaterialSteel, MaterialSteelSD) %>%
  filter(MaterialSteel != 0) %>% arrange(desc(MaterialSteel)) %>%
  mutate(group = ifelse(MaterialSteel > 0, "g1", "g2"))

df_fig_class$taxon_id = factor(df_fig_class$taxon_id, levels = df_fig_class$taxon_id)
  
waterfall_plot_material_class = ggplot(data = df_fig_class, 
           aes(x = taxon_id, y = MaterialSteel, fill = group, color = group)) + 
  geom_bar(stat = "identity", width = 0.7, 
           position = position_dodge(width = 0.4)) +
  geom_errorbar(aes(ymin = MaterialSteel - MaterialSteelSD, ymax = MaterialSteel + MaterialSteelSD), width = 0.2,
                position = position_dodge(0.05), color = "black") + 
  labs(x = NULL, y = "Log fold change" 
       ) + 
  theme_bw() + 
  theme(legend.position = "none", 
        plot.title = element_text(hjust = 0.5),
        panel.grid.minor.y = element_blank(),
        axis.text.x = element_text(angle = 60, hjust = 1, size = 10))

waterfall_plot_material_class
res_material_class$diff_abn

#Genus
phy_ws_23d_filter_genus = aggregate_taxa(phy_ws_23d, "Genus")
phy_ws_23d_filter_genus_filtered = tax_filter(phy_ws_23d_filter_genus, min_prevalence = 1/10)

#Early vs Late

ancom_material_genus = ancombc(phy_ws_23d_filter_genus_filtered, group = "Material", formula = "Material", neg_lb = F)
res_material_genus = ancom_material_genus$res
as.data.frame(res_material_genus)
#Waterfall plot
df_fig1_genus = res_material_genus$lfc %>%
  as.data.frame() %>%
  filter(res_material_genus$diff_abn$MaterialSteel) %>%
  rename(taxon_id = taxon)  # Rename taxon for consistency


df_fig2_genus = data.frame(res_material_genus$se, check.names = FALSE) %>%
  rownames_to_column("taxon_id") %>%
  filter(res_material_genus$diff_abn$MaterialSteel)

colnames(df_fig2_genus)[-1] = paste0(colnames(df_fig2_genus)[-1], "SD")

df_fig_genus = df_fig1_genus %>% left_join(df_fig2_genus, by = "taxon_id") %>%
  transmute(taxon_id, MaterialSteel, MaterialSteelSD) %>%
  filter(MaterialSteel != 0) %>% arrange(desc(MaterialSteel)) %>%
  mutate(group = ifelse(MaterialSteel > 0, "g1", "g2"))

df_fig_genus$taxon_id = factor(df_fig_genus$taxon_id, levels = df_fig_genus$taxon_id)
  
waterfall_plot_material_genus = ggplot(data = df_fig_genus, 
           aes(x = taxon_id, y = MaterialSteel, fill = group, color = group)) + 
  geom_bar(stat = "identity", width = 0.7, 
           position = position_dodge(width = 0.4)) +
  geom_errorbar(aes(ymin = MaterialSteel - MaterialSteelSD, ymax = MaterialSteel + MaterialSteelSD), width = 0.2,
                position = position_dodge(0.05), color = "black") + 
  labs(x = NULL, y = "Log fold change" 
       ) + 
  theme_bw() + 
  theme(legend.position = "none", 
        plot.title = element_text(hjust = 0.5),
        panel.grid.minor.y = element_blank(),
        axis.text.x = element_text(angle = 60, hjust = 1, size = 10))

waterfall_plot_material_genus
res_material_genus$diff_abn

#Figure 5
Figure_5 = waterfall_plot_material_class / waterfall_plot_material_genus + plot_annotation(tag_levels = "A")
Figure_5

ggsave("Figures_all_Material/Figure_5_ancomBC.png", width = 11)
ggsave("Figures_all_Material/Figure_5_ancomBC.tiff", width = 11, dpi = 300)

save_plot("Figures_all_Material/Figure_5_ancomBC_cow.tiff", Figure_5, base_height = 9)
```

